### PR TITLE
feat(http/cookie): add set-cookie headers parser

### PR DESCRIPTION
**Pull request by iuioiua**

### CODE REVIEW

File: http/cookie.ts [hunk 1]
```diff
@@ -218,3 +218,97 @@ export function deleteCookie(
...
+      "Cookies with names starting with `__Host-` must have path be `/`.",
+    );
+  }
+  return cookie;
+}
+
+/**
+ * Parse set-cookies of a header
+ * @param {Headers} headers The headers instance to get set-cookies from
+ * @return {Object} Object with cookie names as keys
```
Comment (kt3k) [inline http/cookie.ts:303]:
```suggestion
 * @return {Cookie[]} List of cookies
```

File: http/cookie.ts [hunk 2]
```diff
@@ -218,3 +218,97 @@ export function deleteCookie(
...
+        }
+        cookie.unparsed.push([key, value].join("="));
+    }
+  }
+  if (cookie.name.startsWith("__Secure-")) {
+    /** This requirement is mentioned in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie but not the RFC. */
+    assert(
+      cookie.secure,
+      "Cookies with names starting with `__Secure-` must be set with the secure flag.",
+    );
```
Comment (kt3k) [inline http/cookie.ts:279]: I feel it's more user-friendly to show this message as warning and skip this 'set-cookie' value. I believe browsers work like that. (The same applies to the below 3 asserts)

What do you thinkg?

File: http/cookie.ts [hunk 3]
```diff
@@ -218,3 +218,97 @@ export function deleteCookie(
...
+  for (const [key, value] of attrs.slice(1)) {
+    switch (key.toLocaleLowerCase()) {
+      case "expires":
+        cookie.expires = new Date(value);
+        break;
+      case "max-age":
+        cookie.maxAge = Number(value);
+        assert(
+          cookie.maxAge >= 0,
+          "Max-Age must be an integer superior or equal to 0",
```
Comment (kt3k) [inline http/cookie.ts:249]: similar to the below comment, how about just showing warning instead of throwing?

**Review by kt3k — COMMENTED** [review]
@iuioiua Sorry for the delay in review. Thanks for your contribution! Left a few comments

**Comment by kt3k** [comment]
@iuioiua Can you also sign the CLA from the above link?

File: http/cookie.ts (hunk 2 above)
Comment (iuioiua) [reply http/cookie.ts:279]: Yeah, good point. Perhaps `console.assert` is more fitting. What do you think?

Comment (kt3k) [reply http/cookie.ts:279]: Agree. `console.assert` works as well.

File: http/README.md [hunk 4]
```diff
@@ -278,3 +278,19 @@ deleteCookie(headers, "deno", { path: "/", domain: "deno.land" });
 ```
 
 > Note: At the moment multiple `Set-Cookie` in a `Response` is not handled.
+
+### getSetCookies
+
+### getCookies
```
Comment (kt3k) [inline http/README.md:284]: This heading looks like a mistake?

File: http/README.md [hunk 5]
```diff
@@ -278,3 +278,19 @@ deleteCookie(headers, "deno", { path: "/", domain: "deno.land" });
 ```
 
 > Note: At the moment multiple `Set-Cookie` in a `Response` is not handled.
+
+### getSetCookies
```
Comment (kt3k) [inline http/README.md:282]: I think this API is for client side handling of cookies (This is something you use when you're interacting server from client side)

Can you describe a little more about it as other methods are for server side handling of cookies, and users might be confused about the purpose of this API.

File: http/README.md (hunk 4 above)
Comment (iuioiua) [reply http/README.md:284]: No, I named it `getSetCookies` deliberately. I understand the possible confusion given both `get` and `set` prefixes in the same function name, but I believe it makes sense once its purpose is made clear.

[Go's cookie.go](https://github.com/golang/go/blob/master/src/net/http/cookie.go) uses the alike `readSetCookies`, which makes sense to me, personally.

Perhaps, `parseSetCookies` might be a better-suited name. What do you think?

File: http/README.md (hunk 5 above)
Comment (iuioiua) [reply http/README.md:282]: Yes, this is for client-side cookie handling. I can't think of any other context in parsing the `Set-Cookie` header that would be needed besides on the client.

Again, [Go's cookie.go](https://github.com/golang/go/blob/master/src/net/http/cookie.go) has the alike `readSetCookies` function, and there's a [comment in the current code](https://github.com/denoland/deno_std/blob/cac7764e3878ac8c2a3c26a1201f3dfcc15e6910/http/cookie.ts#L194-L195) that points out that such a function is needed (if I read it right).

True, it is the only client-side-based function here. But I think it is useful, and related functionality appears to be wanted by some.

File: http/cookie.ts (hunk 2 above)
Comment (iuioiua) [reply http/cookie.ts]: It turns out using `console.assert` might not be workable. My latest change instead uses `console.warn`. I'm unsure whether it's the best way to do it, but at least invalid cookies aren't thrown now. And I'm not sure whether logging to the console is desirable in tests and such... What do you think?

File: http/cookie.ts (hunk 3 above)
Comment (kt3k) [reply http/cookie.ts:248]: @iuioiua How about making this `console.warn` instead of throwing?

File: http/README.md (hunk 4 above)
Comment (kt3k) [reply http/README.md:284]: `getSetCookies` looks fine to me, but there's another subsection with the title of `getCookies` here. Is this intentional?

File: http/cookie.ts [hunk 6]
```diff
@@ -218,3 +217,105 @@ export function deleteCookie(
...
+  if (!headers.has("set-cookie")) {
+    return [];
+  }
+  return [...headers.entries()]
+    .filter(([key]) => key === "set-cookie")
+    .map(([_, value]) => value)
+    /** Parse each `set-cookie` header separately */
+    .map(parseSetCookie)
+    /** Skip empty cookies */
+    .filter(Boolean) as Cookie[];
```
Comment (kt3k) [inline http/cookie.ts:320]: 👍

**Review by kt3k — COMMENTED** [review]
@iuioiua Sorry for the delay in review. The change mostly looks good to me. Added a few comments about small details

File: http/cookie.ts (hunk 3 above)
Comment (iuioiua) [reply http/cookie.ts]: Oh, missed this. Fixed now.

File: http/README.md (hunk 4 above)
Comment (iuioiua) [reply http/README.md]: Wow, I completely misread that! Apologies. Fixed now 😅

**Review by kt3k — APPROVED** [review]
@iuioiua Thanks for updating! LGTM

**Comment by kt3k** [comment]
CI failures look irrelevant to this PR. I'll look into that

**Comment by iuioiua** [comment]
Closes #1977.

**Pull request merged**